test/e2e/persistent_volumes.go

```diff
@@ -22,155 +22,358 @@ import (
+					framework.Failf("AfterEach: pod %v delete ierror: %v", checkPod.Name, err)
+				}
+				checkPod = nil
+			}
 
-		// Create the PersistentVolume and wait for PersistentVolume.Status.Phase to be Available
-		// defer deletion to clean up the PV should the test fail post-creation.
-		framework.Logf("Creating PersistentVolume")
-		pv, err := c.PersistentVolumes().Create(pv)
-		if err != nil {
-			framework.Failf("Create PersistentVolume failed: %v", err)
+			if pvc != nil {
+				// Delete the PersistentVolumeClaim
+				err = c.PersistentVolumeClaims(ns).Delete(pvc.Name)
+				if err != nil && !apierrs.IsNotFound(err) {
+					framework.Failf("AfterEach: delete of PersistentVolumeClaim %v experienced an unexpected error: %v", pvc.Name, err)
+				}
+				pvc = nil
+			}
+			if pv != nil {
+				deletePersistentVolume(c, pv)
+				pv = nil
+			}
 		}
-		defer deletePersistentVolume(c, pv)
-		framework.WaitForPersistentVolumePhase(api.VolumeAvailable, c, pv.Name, 1*time.Second, 20*time.Second)
+	})
 
-		// Create the PersistentVolumeClaim and wait for Bound phase
-		framework.Logf("Creating PersistentVolumeClaim")
-		pvc, err = c.PersistentVolumeClaims(ns).Create(pvc)
-		if err != nil {
-			framework.Failf("Create PersistentVolumeClaim failed: %v", err)
+	// Execute after *all* the tests have run
+	AddCleanupAction(func() {
+		if nfsServerPod != nil && c != nil {
+			nfsServerPodCleanup(c, NFSconfig)
+			nfsServerPod = nil
 		}
-		framework.WaitForPersistentVolumeClaimPhase(api.ClaimBound, c, ns, pvc.Name, 3*time.Second, 300*time.Second)
+	})
 
-		// Wait for PersistentVolume.Status.Phase to be Bound. Can take several minutes.
-		err = framework.WaitForPersistentVolumePhase(api.VolumeBound, c, pv.Name, 3*time.Second, 300*time.Second)
-		if err != nil {
-			framework.Failf("PersistentVolume failed to enter a bound state: %+v", err)
-		}
-		// Check the PersistentVolume.ClaimRef.UID for non-nil value as confirmation of the bound state.
-		framework.Logf("Checking PersistentVolume ClaimRef is non-nil")
-		pv, err = c.PersistentVolumes().Get(pv.Name)
-		if pv.Spec.ClaimRef == nil || len(pv.Spec.ClaimRef.UID) == 0 {
-			pvJson, _ := json.MarshalIndent(pv, "", "  ")
-			framework.Failf("Expected PersistentVolume to be bound, but got nil ClaimRef or UID: %+v", string(pvJson))
-		}
 
-		// Check the PersistentVolumeClaim.Status.Phase for Bound state
-		framework.Logf("Checking PersistentVolumeClaim status is Bound")
-		pvc, err = c.PersistentVolumeClaims(ns).Get(pvc.Name)
-		if pvcPhase := pvc.Status.Phase; pvcPhase != "Bound" {
-			framework.Failf("Expected PersistentVolumeClaim status Bound. Actual:  %+v.  Error: %+v", pvcPhase, err)
-		}
+	// Individual tests follow:
+	//
+	// Create an nfs PV, a claim that matches the PV, a pod that contains the
+	// claim. Verify that the PV and PVC bind correctly and that the pod can
+	// write to the nfs volume.
+	It("should create a PersistentVolume, Claim, and Pod that will test write access of the volume", func() {
```

Comment:
@jeffvance Unless we can show that our fixes correct the flakiness this test cause, I think we still need the [Flaky] string at the end of the It statement.